### DISCUSSION TITLE
Is there a way to profile on a subsection?

### DISCUSSION THREAD

**Discussion by tjdcs** [P403n1x87/austin#219] [General]
Hi, 

One of my favorite things about profiling when I'm working in matlab are the `profile on` and `profile off` to really profile specific sections of code. Is there something similar in the austin ecosystem? I have a case where I'm debugging a code that has a long initialization on the first run but should be faster subsequently, but because a portion of the code is io bound, it's a little bit difficult to actually narrow down the performance without running thousands of subsequent calls. Which takes too long *due to io*

Any thoughts on my situation? 
Thanks!

**Comment by P403n1x87** [comment]
Ah that's an interesting use case. Austin is an out-of-process profiler, so it is not straight-forward to enable/disable it from a Python application. That is not to say that it is impossible though. The [`austin-python`](https://github.com/P403n1x87/austin-python) package gives a Python interface for invoking Austin. A small tweak of the echo example might give you something that you can start/stop in specific places in your code.

Something like [echion](https://github.com/P403n1x87/echion) provides an in-process alternative to Austin. However, this was not designed with this use case in mind, so there currently isn't a dedicated (nice) API to start/stop the profiler programmatically. However, one can work around this with

```python
from echion.bootstrap import start, stop

# Start echion
start()

...

# Stop echion
stop()
```

Annoyingly, the configuration would have to be passed by setting environment variables for now (https://github.com/P403n1x87/echion/blob/c3c522c48cf39757b7cf95922b631663b18b38e2/echion/bootstrap/__init__.py#L33-L37), e.g. to set the output file

```python
import os

os.environ["ECHION_OUTPUT"] = "/path/to/output/file"
```

which would be in the [MOJO](https://github.com/P403n1x87/austin/wiki/The-MOJO-File-Format) binary format (it can be visualised with the [VS Code extension](https://marketplace.visualstudio.com/items?itemName=p403n1x87.austin-vscode)).

**Reply by tjdcs** [reply]
hmm. I really like the echion work around. I've used it before when there was some issues with MacOS 14 that weren't debugged yet. Thanks I will give it a try, if it suits my usage well I'll try write up a little report as I frequently found `profile on` and `profile off` very useful. 

As always, I'm a big fan of these tools. What's the best way to send you a coffee or a few bucks for your help and open source contributions?

**Reply by P403n1x87** [reply]
> As always, I'm a big fan of these tools. What's the best way to send you a coffee or a few bucks for your help and open source contributions?

Ah thanks, that's very much appreciated! I'm glad these tools are useful to others too! If you would like to support these projects, there are a few options (including one-off donations) listed under the *Sponsor this project* section on the right side panel of the [Austin](https://github.com/P403n1x87/austin) project page 🙏
